File: src/main/java/com/telenordigital/prime/ocs/EventHandlerImpl.java
package com.telenordigital.prime.ocs;

import com.lmax.disruptor.EventHandler;
import com.telenordigital.prime.disruptor.PrimeEvent;
import io.grpc.stub.StreamObserver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static com.google.common.base.Preconditions.checkNotNull;

final class EventHandlerImpl implements EventHandler<PrimeEvent> {

    private static final Logger LOG = LoggerFactory.getLogger(EventHandlerImpl.class);

    private final  OcsService ocsService;

    public EventHandlerImpl(final OcsService ocsService) {
        this.ocsService = checkNotNull(ocsService);
    }

    @Override
    public void onEvent(
            final PrimeEvent event,
            final long sequence,
            final boolean endOfBatch) throws Exception {

        try {
            switch (event.getMessageType()) {
                case FETCH_DATA_BUCKET:
                    handleFetchDataBucket(event);
                    break;

                case RETURN_UNUSED_DATA_BUCKET:
                    handleReturnUnusedDataBucket(event);
                    break;

                case TOPUP_DATA_BUNDLE_BALANCE:
                    handleTopupDataBundleBalance(event);
                    break;
                default:
            }
        } catch (Exception e) {
            LOG.warn("Exception handling prime event in OcsService", e);
        }
    }

    private void handleTopupDataBundleBalance(final PrimeEvent event) {
        final ActivateResponse response =
                ActivateResponse.newBuilder().
                        setMsisdn(event.getMsisdn()).
                        build();
        ocsService.activateOnNextResponse(response);
    }

    private void handleReturnUnusedDataBucket(final PrimeEvent event) {
        if (event.getOcsgwStreamId() != null) {
            LOG.info("Returning returnUnusedData response :: for MSISDN: {}", event.getMsisdn());

            final ReturnUnusedDataResponse returnDataInfo =
                    ReturnUnusedDataResponse.newBuilder().
                            setMsisdn(event.getMsisdn()).
                            build();
            final StreamObserver<ReturnUnusedDataResponse> returnUnusedDataResponse
                   =  ocsService.getUnusedDataClientForStream(event.getOcsgwStreamId());
            if (returnUnusedDataResponse != null) {
                returnUnusedDataResponse.onNext(returnDataInfo);
            }
        }
    }

    private void logEventPRocessing(final String msg, final PrimeEvent event) {
        LOG.info("{} :: for MSISDN: {} of {} bytes with request id: {}",
                msg, event.getMsisdn(), event.getBucketBytes(), event.getOcsgwRequestId());
    }

    private void handleFetchDataBucket(final PrimeEvent event) {

        logEventPRocessing("Returning fetchDataBucket response", event);

        try {
            final FetchDataBucketInfo fetchDataInfo =
                    FetchDataBucketInfo.newBuilder().
                            setMsisdn(event.getMsisdn()).
                            setBytes(event.getBucketBytes()).
                            setRequestId(event.getOcsgwRequestId()).
                            build();

            final StreamObserver<FetchDataBucketInfo> fetchDataBucketResponse
            = ocsService.getDataBucketClientForStream(event.getOcsgwStreamId());

            if (fetchDataBucketResponse != null) {
                fetchDataBucketResponse.onNext(fetchDataInfo);
            }
        } catch (Exception e) {
            LOG.warn("Exception handling prime event", e);
            logEventPRocessing("Exception returning fetchDataBucket response", event);
            // unable to send fetchDataBucket response. So, return bucket bytes back to data bundle.
            ocsService.returnUnusedDataBucketEvent(
                    event.getMsisdn(),
                    event.getBucketBytes());
        }
    }
}

File: ocs/src/main/kotlin/org/ostelco/prime/ocs/EventHandlerImpl.kt
package org.ostelco.prime.ocs

import com.lmax.disruptor.EventHandler
import org.ostelco.ocs.api.ActivateResponse
import org.ostelco.ocs.api.CreditControlAnswerInfo
import org.ostelco.ocs.api.FinalUnitAction
import org.ostelco.ocs.api.FinalUnitIndication
import org.ostelco.ocs.api.MultipleServiceCreditControl
import org.ostelco.ocs.api.ReportingReason
import org.ostelco.ocs.api.ServiceUnit
import org.ostelco.prime.disruptor.EventMessageType.CREDIT_CONTROL_REQUEST
import org.ostelco.prime.disruptor.EventMessageType.TOPUP_DATA_BUNDLE_BALANCE
import org.ostelco.prime.disruptor.OcsEvent
import org.ostelco.prime.logger

/**
 * An event handler, handling the [OcsEvent] messages that
 * are used by the Disruptor execution mechanism to handle events.
 */
internal class EventHandlerImpl(private val ocsService: OcsService) : EventHandler<OcsEvent> {

    private val logger by logger()

    override fun onEvent(
            event: OcsEvent,
            sequence: Long,
            endOfBatch: Boolean) {

        try {
            dispatchOnEventType(event)
        } catch (e: Exception) {
            logger.warn("Exception handling prime event in OcsService", e)
        }
    }

    private fun dispatchOnEventType(event: OcsEvent) {
        when (event.messageType) {
            CREDIT_CONTROL_REQUEST -> handleCreditControlRequest(event)
            TOPUP_DATA_BUNDLE_BALANCE -> handleTopupDataBundleBalance(event)

            else -> {} // do nothing
        }
    }

    private fun handleTopupDataBundleBalance(event: OcsEvent) {
        event.msisdnToppedUp?.forEach { msisdn ->
            val response = ActivateResponse.newBuilder().setMsisdn(msisdn).build()
            ocsService.activateOnNextResponse(response)
        }
    }

    private fun logEventProcessing(msg: String, event: OcsEvent) {
        val logString = """
            $msg
            Msisdn: ${event.msisdn}
            Requested bytes: ${event.request?.msccList?.firstOrNull()?.requested?.totalOctets ?: 0L}
            Used bytes: ${event.request?.msccList?.firstOrNull()?.used?.totalOctets ?: 0L}
            Bundle bytes: ${event.bundleBytes}
            Topup bytes: ${event.topUpBytes}
            Request id: ${event.request?.requestId}
        """.trimIndent()

        logger.info(logString)
    }

    private fun handleCreditControlRequest(event: OcsEvent) {

        logEventProcessing("Returning Credit-Control-Answer", event)

        // FIXME martin: This assume we only have one MSCC
        // TODO martin: In case of zero balance we should add appropriate FinalUnitAction

        try {
            val creditControlAnswer = CreditControlAnswerInfo.newBuilder()
                    .setMsisdn(event.msisdn)

            event.request?.let { request ->
                if (request.msccCount > 0) {
                    val msccBuilder = MultipleServiceCreditControl.newBuilder()
                    msccBuilder.setServiceIdentifier(request.getMscc(0).serviceIdentifier)
                            .setRatingGroup(request.getMscc(0).ratingGroup)
                            .setValidityTime(86400)

                    if ((request.getMscc(0).reportingReason != ReportingReason.FINAL) && (request.getMscc(0).requested.totalOctets > 0)) {
                        msccBuilder.granted = ServiceUnit.newBuilder()
                                .setTotalOctets(event.reservedBucketBytes)
                                .build()
                        if (event.reservedBucketBytes < request.getMscc(0).requested.totalOctets) {
                            msccBuilder.finalUnitIndication = FinalUnitIndication.newBuilder()
                                    .setFinalUnitAction(FinalUnitAction.TERMINATE)
                                    .setIsSet(true)
                                    .build()
                        }
                    } else {
                        // Use -1 to indicate no granted service unit should be included in the answer
                        msccBuilder.granted = ServiceUnit.newBuilder()
                                .setTotalOctets(-1)
                                .build()
                    }
                    creditControlAnswer.addMscc(msccBuilder.build())
                }
                creditControlAnswer.setRequestId(request.requestId)
            }

            val streamId = event.ocsgwStreamId
            if (streamId != null) {
                ocsService.sendCreditControlAnswer(streamId, creditControlAnswer.build())
            }
        } catch (e: Exception) {
            logger.warn("Exception handling prime event", e)
            logEventProcessing("Exception sending Credit-Control-Answer", event)

            // unable to send Credit-Control-Answer.
            // So, return reserved bucket bytes back to data bundle.
            ocsService.returnUnusedDataBucketEvent(
                    event.msisdn!!, // TODO vihang: need proper null check
                    event.reservedBucketBytes)
        }

    }
}
